text/0149-exponential-histogram.md

```diff
@@ -0,0 +1,162 @@
+# Add exponential bucketing to histogram protobuf
+
+Add exponential bucketing to histogram protobuf
+
+## Motivation
+
+Currently, the OTEL protobuf [protocol](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto) only supports explicit bound buckets. Each bucket's bound and count must be explicitly defined. This is inefficient to transport buckets whose bounds have a pattern, such as exponential (ie. log scale) buckets. More importantly, without bucket pattern info, the receiver may not be able to optimize its processing on these buckets. With a protocol that also supports exponential buckets, the bounds can be encoded with just a few parameters, regardless of the number of buckets, and the receiver can optimize processing based on the knowledge that these buckets are exponential.
+
+The explicit bucket type will be kept, as a fallback for arbitrary bucket bounds. For example, Prometheus histograms often come with arbitrary user defined bounds.
+
+Exponential buckets are chosen to be added because they are very good at representing [long tail](https://en.wikipedia.org/wiki/Long_tail) distributions, which are common in the OTEL target application like response time measurement. Exponential buckets (ie. log scale buckets) need far fewer buckets than linear scale buckets to cover the wide range of a long tail distribution. Furthermore, percentiles/quantiles can be computed from exponential buckets with constant relative error across the full range.
+
+## Explanation
+
+Exponential buckets will be added. In general, bucket bounds are in the form of:
+
+```
+bound = base ^ exponent
+```
+
+where base is a parameter of a bound series, and exponent is an integer. Note that exponent may be positive, negative, or 0. Such bounds are also commonly known as "log scale bounds".
+
+## Internal details
+
+Proposed message type to be added:
+
+```
+message ExponentialBuckets {
+    double base = 1;
+    double zero_count = 2; // Count of values exactly at zero.
+    ExponentialBucketCounts positive_value_counts = 3;
+    ExponentialBucketCounts negative_value_counts = 4; // Negative values are bucketed with their absolute values
+}
+
+// "repeated double bucket_counts" represents an array of N numbers from bucket_counts[0] to bucket_counts[N-1].
+// With index i starting at 0, ending at N-1, ExponentialBucketCounts defines N buckets, where
+// bucket[i].start_bound = base ^ (i + exponent_offset)   // inclusive bound
+// bucket[i].end_bound = base ^ (i + 1 + exponent_offset) // exclusive bound
```

Comment:
As a note: I believe that removing the bound inclusiveness constraint would in some cases allow or make it easier to optimize the sketch insertion speed at no relative accuracy cost (see [this](https://github.com/DataDog/sketches-java/blob/master/src/main/java/com/datadoghq/sketch/ddsketch/mapping/LogLikeIndexMapping.java#L80) in DDSketch for instance). It may come with some disadvantages (e.g., non-determinism or added complexity in the bucket definition), which we could mitigate maybe, but given the significant performance gain it may allow, I thought it was worth mentioning.

Reply:
In my opinion, the inclusiveness/exclusiveness is irrelavant in practice. It is almost impossible to get consistent boundaries on different platforms anyway. The reason is that the result of the power function is platform/implementation dependent, if the base is not a power of two. A base that is a power of two would imply relative errors in the order of 100% which is rarely useful. If boundaries are implicitly defined, I think there is no other way than to precisely define all numerical operations required to compute the boundaries or - more reasonably - to simply be tolerant with respect to mapping values to buckets.

Reply:
I was on the fence on this one. So unless there is strong objection, I will remove the inclusive/exclusive comment in the next update.

Reply:
Interesting -- I would be happy to see this left unspecified, in order to allow implementations to decide how to treat values that lie on the boundary (subject to precision, ...).